source/server/connection_handler_impl.cc

```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
 void ConnectionHandlerImpl::removeFilterChains(
     uint64_t listener_tag, const std::list<const Network::FilterChain*>& filter_chains,
     std::function<void()> completion) {
-  for (auto& listener : listeners_) {
-    if (listener.second.listener_->listenerTag() == listener_tag) {
-      listener.second.listener_->onFilterChainDraining(filter_chains);
-      break;
-    }
+  auto listener_it = listener_map_by_tag_.find(listener_tag);
+  if (listener_it != listener_map_by_tag_.end()) {
+    listener_it->second->listener_->onFilterChainDraining(filter_chains);
   }
+
   // Reach here if the target listener is found or the target listener was removed by a full
   // listener update. In either case, the completion must be deferred so that any active connection
   // referencing the filter chain can finish prior to deletion.
   Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
 }
 
 void ConnectionHandlerImpl::stopListeners(uint64_t listener_tag) {
-  for (auto& listener : listeners_) {
-    if (listener.second.listener_->listenerTag() == listener_tag) {
-      listener.second.listener_->shutdownListener();
-    }
+  auto iter = listener_map_by_tag_.find(listener_tag);
+  if (iter != listener_map_by_tag_.end()) {
+    iter->second->listener_->shutdownListener();
   }
 }
 
 void ConnectionHandlerImpl::stopListeners() {
-  for (auto& listener : listeners_) {
-    listener.second.listener_->shutdownListener();
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
```

Comment:
when will `listener()` return nullptr? May be keep the original logic is ok?

Reply:
It can be nullptr when you try to stop the listener twice. Although the `shutdownListener` is just going to reset the unique pointer and that is fine, but I think it is better to have a check, seems we can't assume what `shutdownListener` will do in the future.

Reply:
🤔 SGTM.

Reply:
Seems the ActiveTCPListener have the check of listener is nullptr or not for `pauseListener` https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_tcp_listener.cc#L113
but the ActiveUDPListener doesn't do that https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_udp_listener.h#L97